File: 0955. Delete Columns to Make Sorted II/Solution.py
class Solution:
    def minDeletionSize(self, strs: list[str]) -> int:
        n, m = len(strs), len(strs[0])
        ambiguous_rows = list(range(n - 1))

        result = 0
        for j in range(m):
            for i in ambiguous_rows:
                if strs[i][j] > strs[i + 1][j]:
                    result += 1
                    break
            else:
                size = 0
                for i in ambiguous_rows:
                    if strs[i][j] == strs[i + 1][j]:
                        ambiguous_rows[size] = i
                        size += 1
                del ambiguous_rows[size:]
            if not ambiguous_rows:
                break
        return result

File: 0955. Delete Columns to Make Sorted II/Solution.cpp
#include <numeric>
#include <ranges>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int minDeletionSize(vector<string>& strs)
    {
        int m = strs.size(), n = strs[0].size(), result = 0;
        vector<int> prev_sorted(m, 0);
        for (int j = 0; j < n; ++j) {
            bool del = false;
            for (int i = 1; i < m; ++i) {
                if (!prev_sorted[i] && strs[i - 1][j] > strs[i][j]) {
                    ++result;
                    del = true;
                    break;
                }
            }

            if (!del) {
                for (int i = 1; i < m; ++i)
                    prev_sorted[i] = prev_sorted[i] || (strs[i - 1][j] < strs[i][j]);
            }
        }
        return result;
    }
};
